### ISSUE TITLE
Add deflate compression support to the RestAPI

### ISSUE THREAD

**Issue by nicolargo** [nicolargo/glances#1182]
#### Description

For the moment, Glances did not compress data in client/server mode (XML/RPC or REST/JSON). The goal of this enhancement is to (optionally) support Gzip compression for the RestAPI. If the client sent a request with 'Accept-Encoding: gzip' in the Header then Glances should compress the respond and set the 'Content-Encoding' to 'gzip'.

Two actions related to this issue:
* the Glances Restful/JSON server should support the Gzip compression (action @nicolargo)
* the Glances WebUI should be adapted to make request with "Accept-Encoding: gzip" and uncompress the respond (action @notFloran)

### Note for developer

http://www.sami-lehtinen.net/blog/etag-and-gzip-decorators-for-bottlepy

Note: for Python 2 use zlib instead of Gzip lib (https://docs.python.org/2/library/zlib.html#module-zlib)

**Comment by nicolargo** [comment]
First test without compression:

```
# curl -H "Content-Type: application/json" -v http://localhost:61208/api/2/all
> Content-Type: application/json
> 
< Content-Length: 133659
```

and with compression (Gzip):

```
# curl -H "Content-Type: application/json" -H "Accept-encoding: gzip" --compressed -v > Content-Type: > >Content-Type: application/json
> Accept-encoding: gzip
>
< Content-Length: 19704
```
==> Compression: 85%

Had to do a real test (grab all stats every 3 seconds) to check the CPU consumption.

**Comment by nicolargo** [comment]
@notFloran I just push the branch with the implementation on the server side.

Works like a charm with the Curl command (see previous comment), but...

1) the Web UI did not work anymore.. (blocked to the Glances splash screen)
2) the Web UI requests should be change to force the Gzip compression (in the Header request) and uncompress the result returned by the server

Can you have a look ?

**Comment by nicolargo** [comment]
Note for me: have a look on https://github.com/requests/requests/issues/1753#issuecomment-29274517

**Comment by notFloran** [comment]
@nicolargo All the work is made by the browser, I can't do nothing in the webui code to handle the gzip compression.

I think the problem came from the gzip handling in the python code, may be the length of the response is invalid or a header is missing.

Example with httpie: 

```
$ http http://0.0.0.0:61208/api/2/all
HTTP/1.0 200 OK
Access-Control-Allow-Headers: Origin, Accept, Content-Type, X-Requested-With, X-CSRF-Token
Access-Control-Allow-Methods: GET, POST, PUT, OPTIONS
Access-Control-Allow-Origin: *
Content-Encoding: gzip
Content-Length: 29494
Content-Type: application/json
Date: Sat, 11 Nov 2017 10:47:07 GMT
Server: WSGIServer/0.1 Python/2.7.14
Vary: Accept-Encoding

http: error: ContentDecodingError: ('Received response with content-encoding: gzip, but failed to decode it.', error('Error -3 while decompressing data: incorrect header check',))
```

**Comment by nicolargo** [comment]
Strange, it works with Curl but not Httpie...

I will have a look on the server side...

**Comment by nicolargo** [comment]
Solved thanks to Guillaume: https://stackoverflow.com/questions/47240972/httpie-can-not-decode-my-bottle-api-gzipped-respond

### Advanced tests with results

Without compression: 812 Kbps / CPU consumption 2.7% on my laptop (server side)
With compression (default compression level ~ 6): 36 Kbps  / CPU consumption 2.9% on my laptop (server side)

**Comment by nicolargo** [comment]
Ok pushed on the #issue1182 branch.

Current status:
- GZip compression and WebUI OK with Python 2.7
- Internal Server Error with Python 3.5 (no other error displayed, nothing in the log file...)

@asergi if you can have a look on it...

**Comment by nicolargo** [comment]
@asergi Can you have a look on this issue, i want to include this issue in the Glances 3.0 release but i do not found any solution to this Python 3 issue...

**Comment by nicolargo** [comment]
New test, same issue with:

# Python 2.7 and Bottle 0.12.13

Without compression: OK

```
$  curl -H "Content-Type: application/json" -v http://localhost:61208/api/2/all 
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 61208 (#0)
> GET /api/2/all HTTP/1.1
> Host: localhost:61208
> User-Agent: curl/7.47.0
> Accept: */*
> Content-Type: application/json
> 
* HTTP 1.0, assume close after body
< HTTP/1.0 200 OK
< Date: Sat, 24 Feb 2018 15:40:04 GMT
< Server: WSGIServer/0.1 Python/2.7.12
< Content-Length: 112204
< Access-Control-Allow-Headers: Origin, Accept, Content-Type, X-Requested-With, X-CSRF-Token
< Content-Encoding: identity
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Methods: GET, POST, PUT, OPTIONS
< Content-Type: application/json
< 
```

With compression:

```
$ curl -H "Content-Type: application/json" -H "Accept-encoding: gzip" --compressed -v http://localhost:61208/api/2/all
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 61208 (#0)
> GET /api/2/all HTTP/1.1
> Host: localhost:61208
> User-Agent: curl/7.47.0
> Accept: */*
> Content-Type: application/json
> Accept-encoding: gzip
> 
* HTTP 1.0, assume close after body
< HTTP/1.0 200 OK
< Date: Sat, 24 Feb 2018 15:42:39 GMT
< Server: WSGIServer/0.1 Python/2.7.12
< Content-Type: application/json
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Methods: GET, POST, PUT, OPTIONS
< Access-Control-Allow-Headers: Origin, Accept, Content-Type, X-Requested-With, X-CSRF-Token
< Content-Encoding: gzip
```

# Python 3.5 and Bottle 0.12.13

Without compression: OK

```
$ curl -H "Content-Type: application/json" -v http://localhost:61208/api/2/all 
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 61208 (#0)
> GET /api/2/all HTTP/1.1
> Host: localhost:61208
> User-Agent: curl/7.47.0
> Accept: */*
> Content-Type: application/json
> 
* HTTP 1.0, assume close after body
< HTTP/1.0 200 OK
< Date: Sat, 24 Feb 2018 15:43:29 GMT
< Server: WSGIServer/0.2 CPython/3.5.2
< Content-Type: application/json
< Content-Length: 106027
< Content-Encoding: identity
< Access-Control-Allow-Methods: GET, POST, PUT, OPTIONS
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Headers: Origin, Accept, Content-Type, X-Requested-With, X-CSRF-Token
< 
```

With compression: **NOT OK**

```
$ curl -H "Content-Type: application/json" -H "Accept-encoding: gzip" --compressed -v http://localhost:61208/api/2/all
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 61208 (#0)
> GET /api/2/all HTTP/1.1
> Host: localhost:61208
> User-Agent: curl/7.47.0
> Accept: */*
> Content-Type: application/json
> Accept-encoding: gzip
> 
* HTTP 1.0, assume close after body
< HTTP/1.0 500 Internal Server Error
< Date: Sat, 24 Feb 2018 15:43:53 GMT
< Server: WSGIServer/0.2 CPython/3.5.2
< Content-Length: 749
< Content-Type: text/html; charset=UTF-8
< 

    <!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
    <html>
        <head>
            <title>Error: 500 Internal Server Error</title>
            <style type="text/css">
              html {background-color: #eee; font-family: sans;}
              body {background-color: #fff; border: 1px solid #ddd;
                    padding: 15px; margin: 15px;}
              pre {background-color: #eee; border: 1px solid #ddd; padding: 5px;}
            </style>
        </head>
        <body>
            <h1>Error: 500 Internal Server Error</h1>
            <p>Sorry, the requested URL <tt>&#039;http://localhost:61208/api/2/all&#039;</tt>
               caused an error:</p>
            <pre>Unhandled exception</pre>
        </body>
    </html>
* Closing connection 0
```

**Comment by notFloran** [comment]
With [this stackoveflow question](https://stackoverflow.com/questions/1835585/python-zlib-not-decodable-when-returned-by-an-http-response) I found this solution :  in `glances_bottle.py` replace `response.headers['Content-Encoding'] = 'gzip'` with `response.headers['Content-Encoding'] = 'deflate'`.

No more `ERR_CONTENT_DECODING_FAILED` and the WebUI is now Ok

**Comment by nicolargo** [comment]
Perfect ! Thanks @notFloran 

Merged on the develop branch.

Will be available in Glances 3.0 :)
